File: Zipfsong/Zipfsong/Zipfsong.java
import java.util.*;
import java.lang.Long;

public class Zipfsong {

	public List<Song> songList;
	private long select;

	private class Song implements Comparable<Song> {

		public long zi;
		public String title;

		private Song(long zi, String title) {
			this.zi = zi;
			this.title = title;
		}

		public int compareTo(Song a) {
			if(zi < a.zi)
				return 1;
			else if (zi == a.zi)
				return 0;
			else
				return -1;
		}
	}

	// Add a song to the linked list
	public void addSong(int position, long plays, String title) {

		long zi = plays*position;
		Song song = new Song(zi, title);

		if(songList.size() == 0)
			songList.add(song);
		else {
			for(int i = 0; i < songList.size(); i++) {
				if(songList.get(i).zi < song.zi) {
					songList.add(i, song);
					break;
				}
				else if(songList.size() < select && i == (songList.size() - 1)) {
					songList.add(song);
					break;
				}
			}
		}
		
		while(songList.size() > select)
			songList.remove(songList.size() - 1);
		
	}

	public void displaySongs() {
		for (int i = 0; i < songList.size(); i++) {
			System.out.println(songList.get(i).title);
		}
	}

	public static void main(String[] args) {

		Kattio in = new Kattio(System.in);

		// Get initial variables
		long n = in.getLong();
		long select = in.getLong();
		
		Zipfsong zipf = new Zipfsong();
		zipf.songList = new ArrayList<Song>();
		zipf.select = select;
		
		for(int i = 1; i <= n; i++) {

			// Get the pair of values
			long plays = in.getLong();
			String title = in.getWord();

			zipf.addSong(i, plays, title); 
		
		}

		zipf.displaySongs();
		
	}
}

File: Zipfsong/Zipfsong/Zipfsong.cpp
//
//  main.cpp
//  Zipfsong
//
//  Created by Ben Eisner on 12/29/13.
//  Copyright (c) 2013 Ben Eisner. All rights reserved.
//

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class Song
{
public:
    long zi;
    string title;
    
    Song(int position, long plays, string title)
    {
        this->title = title;
        this->zi = plays*position;
    }
    
    bool operator<(Song otherSong) const
    {
        return zi > otherSong.zi;
    }

    
};

class SongList
{
private:
    vector<Song> songVector;
    long dispNum;
    
public:
    SongList(long dispNum) {
        this->dispNum = dispNum;
    }
    
    void addSong(Song song) {
        songVector.push_back(song);
        stable_sort(songVector.begin(), songVector.end());
        
        if(songVector.size() > dispNum)
            songVector.pop_back();
        
    }
    
    void displaySongs() {
        for(int i = 0; i < dispNum; i++) {
            cout << songVector[i].title << endl;
        }
    }
};

int main(int argc, const char * argv[])
{
    long tracks;
    long displayNum;
    
    cin >> tracks;
    cin >> displayNum;
    
    SongList songList(displayNum);

    
    for(int i = 1; i <= tracks; i++) {
        long plays;
        string title;
        
        cin >> plays;
        cin >> title;
        
        
        Song song(i, plays, title);
        songList.addSong(song);
    }
    
    songList.displaySongs();
    
    
}

